library/utility/signal_handler.hh

```diff
@@ -64,31 +66,70 @@ namespace scarab
+     goes out of scope, signals are still handled until the second instance is destructed.
+
+     In addition to this typical behavior, `signal_handler` includes an interface for more fine-grained control over when 
+     signals are handled while the `signal_handler` exists (i.e. handle_signals(), unhnadle_signals(), and is_handling()), 
+     but the recommended use case is as described above, where signal handling is controled by the existence of a `signal_handler` instance.
+
+     Note that the interface for `signal_handler` uses all static functions, so there is no need to create an extra instance of 
+     the class to use any part of that interface.  So the creation and destruction of instances can be used to control 
+     whether or not signals are handled, and the rest of the interface can be used without an instance.
+
+     # Cancelables
+
+     This class communicates the need to exit to other objects via the \ref cancelable interface.  Simply put, a \ref cancelable object 
+     can be canceled, or told to exit, asynchronously by the `signal_handler`.  Only the cancelable objects that the user wants 
+     to be directly canceled by `signal_handler` should be added with add_cancelable().  In some cases that might be all of the 
+     cancelable objects.  In other cases, where there's some hierarchy of cancelables, it would only be the top-level cancelables.
+
      While add_cancelable() takes a std::shared_ptr<cancelable> as an argument, it stores only std::weak_ptr<cancelable>, and 
      claims no ownership or responsibility for the lifetime of the cancelable object.
+
+     # Manual Cancelation/Exiting/Terminating
+
+     In addition to acting automatically via handled signals, a user can manually terminate, exit, or cancel using 
+     terminate(), exit(), or cancel_all(), respectively.  cancel_all() will only affect the canceled objects.  
+     exit() perform a clean exit using cancel_all(), in addition to setting a return code, though it assumes the 
+     application will take care of actually exiting.  And terminate() will immediately exit the application using std::_Exit().
     */
     class SCARAB_API signal_handler
     {
         public:
             signal_handler();
             virtual ~signal_handler();
 
-            /// Add a cancelable object
-            void add_cancelable( std::shared_ptr< cancelable > a_cancelable );
-            /// Remove a cancelable object
-            void remove_cancelable( std::shared_ptr< cancelable > a_cancelable );
-            /// Remove a cancelable object with a plain pointer
-            void remove_cancelable( cancelable* a_cancelable );
-
-            /// Remove all cancelables and signal handling
-            void reset();
-
             /// Static version: add a cancelable object
-            static void add_cancelable_s( std::shared_ptr< cancelable > a_cancelable );
+            static void add_cancelable( std::shared_ptr< cancelable > a_cancelable );
```

Comment:
All these functions here should probably use raw pointers since they only want to access the resource, right?

See https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-smart

Reply:
This is a really good point.  I started a discussion over in #92 to talk about this, and we can implement a change in a subsequent hotfix.  Thanks for the insight.